src/callbacks_step/analysis.jl

```diff
@@ -195,10 +199,33 @@ function (analysis_callback::AnalysisCallback)(integrator)
   @unpack dt, t = integrator
   iter = integrator.destats.naccept
 
+  # Record performance measurements and compute performance index (PID)
+  runtime_since_last_analysis = 1.0e-9 * (time_ns() - analysis_callback.start_time_last_analysis)
+  # PID is an MPI-aware measure of how much time per global degree of freedom (i.e., over all ranks)
+  # and per `rhs!` evaluation is required. MPI-aware means that it essentially adds up the time
+  # spent on each MPI rank. Thus, in an ideally parallelized program, the PID should be constant
+  # independent of the number of MPI ranks used, since, e.g., using 4x the number of ranks should
+  # divide the runtime on each rank by 4. See also the Trixi.jl docs ("Performance" section) for
+  # more information.
+  # The PID is meant to measure everything except the time required for running the `AnalysisCallback`
+  # itself. However, we call the right-hand side once below to be able to compute quantities such as
+  # the semidiscrete rate of change of the entropy. Thus, the total number of right-hand side calls is
+  # one bigger than the number of right-hand side calls included in the `runtime_since_last_analysis`.
+  # Thus, we need to normalize it appropriately. Checking whether the number of calls is positive
+  # ensures that we get `NaN` instead of a negative number for the first output before any right-hand
+  # evaluations have been performed.
+  ncalls_ = ncalls(semi.performance_counter)
+  ncalls_ = ifelse(ncalls_ > 0, ncalls_, ncalls_ - 1)
```

Comment:
@ranocha Is this code really robust to produce the correct number of `rhs!`? Alternatively, we could just store the `ncalls` counter at the end of the analysis CB and then use the difference to that. It would be uglier, but maybe easier to verify that it is the correct number?

Reply:
Sounds good :+1:

Reply:
We should just make sure that we exclude the RHS call in line 285 or so

Reply:
I attempted a fix in 492c1996. however, the first count is still one too high - any idea why?

Reply:
Yes:
https://github.com/SciML/OrdinaryDiffEq.jl/blob/5ca6cd6dc8371ac3009c3af46c24ab8a4421f41d/src/perform_step/low_storage_rk_perform_step.jl#L44

This is needed for 
- continuous callbacks
- dense output
- maybe something else I don't remember right now

We could try to figure our whether we can add a bunch of checks whether any of these features are used an remove this additional RHS call in OrdinaryDiffEq.jl - but it does not seem to be an extremely urgent problem to me.

Reply:
You could try to use something along the lines of

```julia
intergrator = init(#= same arguments as for solve =#)
step!(integrator)
# call analysis callback using our interface to reset stuff
# something to solve to the end without further interruption
```

Reply:
I am just thinking - is this even an issue? I mean: If we reset both *counter* and *timer* in the initialization of the callback, then the callback is invoked once on the IC, and then afterwards we get one additional `rhs!` evaluation, we do get a consistent measurement result, don't we?

The only thing might be that on top of that extra `rhs!` call, OrdinaryDiffEq might to other non-negligible initialization stuff that gets folded into the *timer* but is not accounted for in the *counter*, but for a performance measurement, either the number of time steps or the workload per step should be large enough to not make this count.

Either way, I will include your example (with a fix: it should read
```
intergrator = init(#= same arguments as for solve =#)
step!(integrator)
analysis_callback.affect!(integrator)
solve!(integrator) # to solve to the end without further interruption
```
i.e., we need to further unpack the actual callback from the `DiscreteCallback`)
in the docs and re-request a review.

Reply:
Done in 2aee37eb. In fact, I did not include your example, since it is also not a perfect fit. For example, it will not take into account the time spent in all callbacks that usually occur *after* the analysis callback is called, which is, among other callbacks, I/O, AMR, or CFL-based stepsize. Since these callbacks - should they be enabled - actually do have a measurable impact on the performance (while the initialization of `OrdinaryDiffEq` seems benign besides the `rhs!` call), I think the solution of manually calling the callback is just a bandaid that potentially verschlimmbessert the whole situation.

Reply:
Sounds good to me.